### DISCUSSION TITLE
What if… A player gets injured?

### DISCUSSION THREAD

**Discussion by Sakr9** [kyleskom/NBA-Machine-Learning-Sports-Betting#68] [Ideas]
Hello and first of all congratulations for your work because it is among the most intuitive and simple to use. I am very passionate about statistics and the NBA but I have zero knowledge regarding Python and machine learning and my work has always been limited to using Excel, where I still achieved about 40-45% of correct results, but working on statistics of each player. The ideas I would like to suggest are many:

- Why not try applying your own machine learning to each player's stats? Assists, rebounds and points, since they are easily available data.
- To optimize the success rate of your machine learning, I've obviously seen these methods that are limited to predicting everything based on the team's performance over time but no one implements the initial lineups and the players who will play that match. For example, if GSW is strong because Curry is there then surely it is not possible to predict by machine learning that they are likely to go down around the time he was injured, as it was because… It doesn't know! And that's why I rely on projected minutes of probable lineups that are easily accessible on the web, even in csv format. There are datasets on the web that have lineup history.

An idea could be to analyze which players have played more together, analyze how many points they scored and how the team behaved when one or more players were missing. Therefore, calculate the offensive and defensive strength of the teams when there are those specific players on the field.

- At a later stage, it would also be useful to know how teams behave based on player roles. For example, if there are teams that easily allow 3-point shots, if there are teams that easily allow rebounds under the basket, etc.

I know it's maybe a utopia but never say never. Thank you!

**Comment by kyleskom** [comment]
Hi @Keryfia you are 100% right. It becomes inaccurate when player injuries start to show up. This has been an issue iv thought a lot about and would like to implement. Including player data would be a lot of work / time and I just don't have it currently. Im kinda looking for help with this and others to split this work up. 

As for now using this is just another tool to help. I have to check injuries as well and manually see what outputs make sense for what games. 

With all that being said I hope one day to include player data in calculations.

**Reply by Sakr9** [reply]
> Hi @Keryfia you are 100% right. It becomes inaccurate when player injuries start to show up. This has been an issue iv thought a lot about and would like to implement. Including player data would be a lot of work / time and I just don't have it currently. Im kinda looking for help with this and others to split this work up.
> 
> As for now using this is just another tool to help. I have to check injuries as well and manually see what outputs make sense for what games.
> 
> With all that being said I hope one day to include player data in calculations.

Hello and thanks for the reply. Sure, it's quite a complicated job to implement, I realize!

Among other things, another curiosity is why the Under / Over are always less precise in any neural model or simple statistical model that I have seen. This could be greatly improved by also including data on the distance between cities and the time of days that separate one game from another. For example, doing a comparison right now, I'm seeing that BRK is struggling a lot against DET right now. BRK will probably still win in the end but most likely not by the large margin we could have expected. What's the reason? BRK are more tired because they faced a difficult game against PHI yesterday, while DET played three days ago, so they are fresher. Why was I talking about the distance between cities? Because very often it happens that teams have to play immediately after facing long journeys by plane, sometimes even in less than 24 hours and it happens quite often that if a team wins a game, it loses the next one if it plays in such a short distance of time or vice versa .

In short, they seem like many small useless things but could improve a neural network to "think" better.

**Reply by kyleskom** [reply]
I prefer the XGB models for the game totals but yet again all things I have thought about but really just came down to time. I think cleaner data (not in excel sheets) and data stored properly in a database would also make these different data science tests / experiments easier.

I think the data side needs a massive overhaul and cleanup to make it easier to do for the fun data science and machine learning stuff!

**Reply by kyleskom** [reply]
I would also like to stress that this originally started as a project to just learn basic machine learning. When starting this I didn't care about how the data was and put aside doing things the right way to get to the machine learning part as fast as possible. This project kinda blew up in a way I never thought and now requires a lot more then I was in for. 

Like I said I'm willing to take this project to the next level and work on all those things, just when time allows.

**Reply by Sakr9** [reply]
Congratulations for all the work then, you can see that there has been a lot of work behind it over time.  I also hope to be able to learn the fundamentals of machine learning in my free time because it is a field that attracts me a lot and I find it very interesting, I deal with medicine now in my life but I have a past as a student of electronics and programming and I would like to implement the two things to be able to create something interesting at work for scientific research.  I say this because I can tell you that your project is spurring me to start studying Python and the basics of neural networks again, so I hope that in my small way this can make you proud, also to all the collaborators behind it.  I love clean, easy, and intuitive end-user code, and I hope I can help out with your code one day!  Many greetings, we hope that for the next NBA season we will be able to have an electronic Nostradamus!

**Comment by nealmick** [comment]
Hello Kyle, I recently discovered your project and it is very similar to mine. A few years ago, I started working on a project to predict NBA game winners using TensorFlow. I had no prior knowledge of machine learning, but I gradually learned as I went along. I believe I could improve my project by incorporating some of your ideas. My project outputs a single sigmoid value that ranges from 0-1, indicating the likelihood of the home or visitor team winning. I have trained multiple models using different configurations of input features, mainly focusing on the top players by playtime. Each player's features consist of their season averages. I used an API called [balldontlie.io](http://balldontlie.io/) to collect data and also created PKL binary files with a lot of stats that I can quickly load or reform. The data includes games, team rosters, and player season averages from 2006-19.  I also have several datasets with over 20k games and up to about 250 input features per game. To control the model I built a Django web app which can make predictions, track accuracy stats, and display today's games (https://nbadata.cloud/). The best ive been able to achieve is around 65-70% correct so far.  In the future, I want to refactor my model to target the spread and specific game scores instead of just the winners. The input features would be similar to yours, plus the features I already have. Thank you for all your hard work on this project. Hope to be in touch, and will link my repo below, though it may not be as well organized as yours :)


https://github.com/nealmick/bb

Cheers,
Neal Mick

**Reply by kyleskom** [reply]
Thanks for sharing will take a look!

**Reply by WillC53** [reply]
Hey Neal Mick, I'm interested in developing my own model as well. I'm currently a business analytics student, but unfortunately do not have much experience with machine learning and regression based analytics (yet). As someone who learned over time, how did you go about teaching yourself? What books/resources did you read through to gain an understanding and knowledge base? Thanks!

Will

**Reply by nealmick** [reply]
I recommend the TensorFlow docs.  They have a bunch of tutorial projects, everything from classifying cats and dogs, to types of flowers, even predicting the weather.  I recommend doing a few of those projects and really trying to understand whats going on.  I think the most important part is just learning how to form and restructure your data sets.  I recommend this project here:
https://www.tensorflow.org/quantum/tutorials/mnist
It's a hand written number classifier.  I built the same project with PyTorch and made a web interface for it.  Still one of my favorite projects.  live here https://nealmick.com/nc/ and github here https://github.com/nealmick/NumberClassifier

Overall I would say just jump in, and see where it takes you.  I really recommend having fun with it and taking time to enjoy it too... other wise you will quit.  A big thing that helps me with this is building website interfaces for my projects.  This way you can have fun and mess with the projects and they seem soo much more real, at least compared to command line apps.
